### DISCUSSION TITLE
How to set an HttpRequest in ServerRequestContext in Micronaut 4

### DISCUSSION THREAD

**Discussion by francesco-monorchio** [micronaut-projects/micronaut-core#9606] [General]
Hi all,

In Micronaut 3 I had a Spock Extension that set a mock request in the ServerRequestContext using ServerRequestContext.set() method. From Micronaut 4 the set method does not exist anymore and using ServerRequestContext.with(request, Thread.currentThread) does not work as expected. Is there a way to do this in another way?

Thanks,
Frank

**Comment by yawkat** [comment]
you need to use it like `ServerRequestContext.with(request, () -> { ... code that uses the request ... })`

**Reply by francesco-monorchio** [reply]
Your sample code does not work! I have found this, but I don't know if it is the right approach:

```
class MockRequestInterceptor implements IMethodInterceptor {

    HttpRequest<?> request

    @Override
    void intercept(IMethodInvocation invocation) {

        def context = PropagatedContext.getOrEmpty()
        def element = new ServerHttpRequestContext(request)

        //noinspection ChangeToOperator
        context.plus(element)
                .propagate()

        invocation.proceed()

        context.propagate()
    }

}
```

This is a Spock Interceptor around each feature method!

**Reply by graemerocher** [reply]
this looks like the correct approach /cc @dstepanov

**Reply by dstepanov** [reply]
The solution by Jonas should work. Please show the use-case that doesn't work.
